File: 백준/Silver/14465. 소가 길을 건너간 이유 5/소가 길을 건너간 이유 5.java
import java.io.*;
import java.util.*;

public class Main {
    private static int N, K, B;
    private static List<Integer> nums;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer input = new StringTokenizer(br.readLine());
        N = Integer.parseInt(input.nextToken());
        K = Integer.parseInt(input.nextToken());
        B = Integer.parseInt(input.nextToken());

        nums = new ArrayList<>();
        for(int i = 0; i <= N; i++){
            nums.add(i);
        }
        List<Integer> broken = new ArrayList<>();
        for(int i = 0; i < B ; i++){
            broken.add(Integer.parseInt(br.readLine()));
        }

        System.out.println(f(broken));
    }

    private static int f(List<Integer> broken){
        int ans = 100001;
        int cnt = 0;
        int l = 1, r = l + K -1;
        for(int i = l; i <= r; i++){
            if(broken.contains(i)){
                cnt++;
            }
        }
        while(r < N){
            if(broken.contains(nums.get(l))){
                cnt--;
            }
            l++;
            r++;
            if(broken.contains(nums.get(r))){
                cnt++;
            }
            ans = Math.min(ans, cnt);
        }
        return ans;
    }

}

File: 백준/Silver/14465. 소가 길을 건너간 이유 5/소가 길을 건너간 이유 5.py
n, k, b = map(int, input().split())
broken = set([int(input()) for _ in range(b)])
nums = list(range(n + 1))

# 0 1 2 3 4 5 6 7 8 9 10
#   # #     #       #  #
# l = 1, r = 4
# 한 옆으로 움직이면 기존의 l은 빠지고 새로운 r이 들어온다.
# 새로운 l은 이미 기존의 범위에 포함이 되었기 때문에 고려할 필요 없고
# 기존 r은 이미 기존의 범위에 포함이 되었기 때문에 고려할 필요 없다.

cnt = 0
l = 1
r = l + k - 1
for x in broken:
    if l <= x <= r:
        cnt += 1

ans = 100001
while r < n:
    if nums[l] in broken:
        cnt -= 1
    l += 1
    r += 1
    if nums[r] in broken:
        cnt += 1
    ans = min(ans, cnt)
    
print(ans)